### ISSUE TITLE
[Bug]: Sweeps not working together with PyTorch

### ISSUE THREAD

**Issue by Leon0402** [wandb/wandb#9924]
### Describe the bug

It seem that importing pytorch break wandb sweeps. 

Small example
```
# train.py
import random
import time

# import torch
import tqdm

import wandb


def train_one_epoch(epoch, lr, bs):
    acc = 0.25 + ((epoch / 30) + (random.random() / 10))
    loss = 0.2 + (1 - ((epoch - 1) / 10 + random.random() / 5))
    return acc, loss


def evaluate_one_epoch(epoch):
    acc = 0.1 + ((epoch / 20) + (random.random() / 10))
    loss = 0.25 + (1 - ((epoch - 1) / 10 + random.random() / 6))
    return acc, loss


def main():
    wandb.init()

    lr = wandb.config.lr
    bs = wandb.config.batch_size
    epochs = wandb.config.epochs

    for epoch in tqdm.tqdm(range(epochs)):
        train_acc, train_loss = train_one_epoch(epoch, lr, bs)
        val_acc, val_loss = evaluate_one_epoch(epoch)
        time.sleep(1)

        wandb.log(
            {
                "epoch": epoch,
                "train_acc": train_acc,
                "train_loss": train_loss,
                "val_acc": val_acc,
                "val_loss": val_loss,
            }
        )


main()
```
# config.yaml
program: train.py
method: random
name: sweep
metric:
  goal: maximize
  name: val_acc
parameters:
  batch_size:
    values: [16, 32, 64]
  lr:
    min: 0.0001
    max: 0.1
  epochs:
    values: [5, 10, 15]
```

Start sweep:
 wandb sweep --project sweep-demo-cli config.yaml

And then start agent:
wandb agent <id>


Observe that it is working:
```
wandb: WARNING Using legacy-service, which is deprecated. If this is unintentional, you can fix it by ensuring you do not call `wandb.require('legacy-service')` and do not set the WANDB_X_REQUIRE_LEGACY_SERVICE environment variable.
wandb: Starting wandb agent 🕵️
2025-05-28 22:47:16,585 - wandb.wandb_agent - INFO - Running runs: []
2025-05-28 22:47:16,837 - wandb.wandb_agent - INFO - Agent received command: run
2025-05-28 22:47:16,838 - wandb.wandb_agent - INFO - Agent starting run with config:
        batch_size: 32
        epochs: 15
        lr: 0.01778725577107571
2025-05-28 22:47:16,840 - wandb.wandb_agent - INFO - About to run command: /usr/bin/env python train.py --batch_size=32 --epochs=15 --lr=0.01778725577107571
wandb: WARNING Using legacy-service, which is deprecated. If this is unintentional, you can fix it by ensuring you do not call `wandb.require('legacy-service')` and do not set the WANDB_X_REQUIRE_LEGACY_SERVICE environment variable.
<Here all the output of the script>
...
```


But if you comment in the pytorch import in the training script, suddently the output is:
```
wandb: WARNING Using legacy-service, which is deprecated. If this is unintentional, you can fix it by ensuring you do not call `wandb.require('legacy-service')` and do not set the WANDB_X_REQUIRE_LEGACY_SERVICE environment variable.
wandb: Starting wandb agent 🕵️
2025-05-28 22:52:16,774 - wandb.wandb_agent - INFO - Running runs: []
2025-05-28 22:52:17,476 - wandb.wandb_agent - INFO - Agent received command: run
2025-05-28 22:52:17,476 - wandb.wandb_agent - INFO - Agent starting run with config:
        batch_size: 16
        epochs: 15
        lr: 0.0013364305378862206
2025-05-28 22:52:17,478 - wandb.wandb_agent - INFO - About to run command: /usr/bin/env python train.py --batch_size=16 --epochs=15 --lr=0.0013364305378862206
2025-05-28 22:52:22,491 - wandb.wandb_agent - INFO - Running runs: ['440mqklf']
```
And it will hang forever without any progress. Not great :(

**Comment by paulosabile-wb** [comment]
Hi @Leon0402 Good day and thanks for reporting this to us. We're sorry to hear that you are experiencing issues with your experiments. Could you please share your current SDK version so I can reproduce this issue and further investigate? You can get this by running `wandb --version`. Thanks!

**Comment by Leon0402** [comment]
@paulosabile-wb Happens with 0.19.11, but also earlier version (I had 0.19.5 previously I think). The issue is resolved when downgrading to python 3.12. But obviously that is only a quick fix, not a permanent solution.

**Comment by Leon0402** [comment]
@paulosabile-wb Hey! Is there any progress on reproducing / investigating the issue? Thanks! :)

**Comment by fmamberti-wandb** [comment]
Hi @Leon0402 , thank you for surfacing this and sharing so many details to reproduce the issue. I was able to reproduce this on our end and I have raised it with our engineering team for further investigation and we will keep you posted with any  updates.

While we look into this, I would suggest setting up an environment with python 3.12 in order to run your Sweeps as a workaround.

**Comment by fmamberti-wandb** [comment]
Hi @Leon0402 , I wanted to follow up here and let you know that we narrowed down this to an issue that is not wandb related - this is torch importing to fail if all the following conditions are met:

- Use miniconda (not venv or pyenv)
- Python 3.13 (not 3.12)
- Set preexec_fn=os.setpgrp before calling subprocess.Popen
- Child Python script calls import torch

You can see a wandb-indepent repro on this [repo](https://github.com/kelu-wandb/public-demo/blob/main/torch-setpgrp-conda/torch-setpgrp-test.py).

So another workaround if you need to upgrade to python3.13 would be to use a venv or pyenv environment rather than a conda one.

FYI, I've also raised this as an issue on pytorch repo - you can find it [here](https://github.com/pytorch/pytorch/issues/159645)

**Comment by Leon0402** [comment]
Hi @fmamberti-wandb,

thank you for letting me know. Unfortunately, I am not using (mini)conda, but uv, which should use itself virtualenv I think. The third condition at leas I didn;t do explicityly. Is this something wandb does?

**Comment by valterschutz** [comment]
I have the same issue! I'm using uv as well.

- python 3.13.5
- wandb 0.21.1
- torch 2.8.0

**Comment by fmamberti-wandb** [comment]
Hi @Leon0402 @valterschutz , we have a [PR merged](https://github.com/wandb/wandb/pull/10489/) into main to fix this issue which will be part of next SDK release.
